### ISSUE TITLE
Support composer packages in repositories with type path

### ISSUE THREAD

**Issue by bburnichon** [renovatebot/renovate#11674]
### What would you like Renovate to be able to do?

I have some end-projects with embedded private PHP packages in a project.

These private packages are declared in `composer.json` file in `repositories` section. https://getcomposer.org/doc/05-repositories.md#path.

The issue can be seen when using `*` or `?` in parse using glob patterns of PHP. https://php.net/glob

### If you have any ideas on how this should be implemented, please tell us here.

When a `*` or `?` is detected in `url` field, special care is needed to expand directories and add individual repositories.

Sadly, I am no expert in typescript and can not help with this. But I can help with the PHP way of doing

### Is this a feature you are interested in implementing yourself?

No

**Comment by bburnichon** [comment]
A minimal reproduction repository can be found [here](https://github.com/bburnichon/renovate-composer-path)

**Comment by viceice** [comment]
Can you explain what renovate should do there?

**Comment by bburnichon** [comment]
When declaring a repository with type `path`, it should add packages declared in given path (expanded like php's glob function would).

In test repo above, 
`{"type": "path", "url": "subrepos/*"}` declares package in https://github.com/bburnichon/renovate-composer-path/blob/main/subrepos/subrepo/composer.json (path `subrepos/subrepo` matches) and as such, `bburnichon/subrepo` package could be resolved (an optional version could also be given see [composer doc](https://getcomposer.org/doc/05-repositories.md#path))

**Comment by viceice** [comment]
And what should renovate update? You don't make clear what renovate should do beside detecting that repo.

What should renovate do with that detected dependency?

**Comment by rarkins** [comment]
> Currently, these packages are not detected and look-up is performed on packagist.org which eventually fails

Do you mean they are detected, but they are *misunderstood* which leads to an unnecessary/undesirable lookup from Packagist?

> in case subdirectory is itself a git repository

Do you mean a git submodule? In this case updating a git submodule could/should be done independently of composer. Or is it different - in which case I'd like to hear more details. For example I'm not sure where `{"bburnichon/subrepo": "2.0"}` comes from and what that `2.0` refers to

**Comment by bburnichon** [comment]
> > Currently, these packages are not detected and look-up is performed on packagist.org which eventually fails
> 
> Do you mean they are detected, but they are _misunderstood_ which leads to an unnecessary/undesirable lookup from Packagist?
> 
They are not detected at all in path location. An undesirable lookup from Packagist is performed. If packagist had a package with the same name, troubles could occur.

> > in case subdirectory is itself a git repository
> 
> Do you mean a git submodule? In this case updating a git submodule could/should be done independently of composer. Or is it different - in which case I'd like to hear more details. For example I'm not sure where `{"bburnichon/subrepo": "2.0"}` comes from and what that `2.0` refers to

Could be a submodule, but nothing says it should belong to the same repository as the project. A special docker image for php could be used with several repositories cloned in a well known path (`/opt/lib/php/private-packages/...` for example)

What I tried to explain is 2 composer repositories are declared:
```
{"type": "path", "url": "subrepos/subrepo-v1", "versions": {"bburnichon/subrepo": "1.0"}},
{"type": "path", "url": "subrepos/subrepo-v2", "versions": {"bburnichon/subrepo": "2.0"}},
```

in path `subrepos/subrepo-v1`, a submodule could be checkout out with a library code in version `1.0`
in path `subrepos/subrepo-v2`, a submodule could be checkout out with the same library but in version `2.0`

renovatebot should be able to indicate migration of root project to use package `bburnichon/subrepo` with version 2.0 without resorting to call packagist.

I will try to change my reproduction repository to something mimicing this.

**Comment by rarkins** [comment]
I think the best thing we can do to start is to ensure we skip such dependencies instead of looking them up from Packagist. After that, we can look into which scenarios we *could* update. e.g. if it's outside the project like `/opt/lib` then we definitely can't.

**Comment by bburnichon** [comment]
Yes, the simpler case is just to ensure this kind of package have the right hash in `composer.lock`.
Trying to setup a more advanced case is premature I think. And composer currently prevents such cases.

With
https://github.com/bburnichon/renovate-composer-path/commit/54a5bc799bfc45db165eb91abcd9a94c182f2add

```
$ composer2 update
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Root composer.json requires bburnichon/full-path ^1.0, it is satisfiable by bburnichon/full-path[1.1] from path repo (full/path1.1) but bburnichon/full-path[2.0] from path repo (full/path2) has higher repository priority. The packages with higher priority do not match your constraint and are therefore not installable. See https://getcomposer.org/repoprio for details and assistance.
```

**Comment by rarkins** [comment]
I wrote "ensure we skip such dependencies instead of looking them up from Packagist" and you appeared to agree with "Yes" but then you wrote "ensure this kind of package have the right hash in composer.lock." which does not sound like what I wrote.

**Comment by bburnichon** [comment]
You're right, I did agree with skipping as the first step. When answering I thought about the lock thing.

So first step is "ensure we skip such dependencies instead of looking them up from Packagist"
Then, it would be nice to ensure hash in composer.lock, if present, is up-to-date.
Then, full dependency lookup could be done, with the non-canonical repository option that I just discovered.
I mean, I do have cases with submodules installed in a big project but no multi-versions cases currently.
To ensure that these submodules are up-to-date would be nice but I can live without it.

**Comment by rarkins** [comment]
Let's do this:
- We need a minimal reproduction repo for a scenario where a path dependency is being looked up from Packagist incorrectly. Presumably it can be as small as one dependency?
- We mark this issue as status:ready 
- Update code to ensure it skips such dependencies and close this issue
- Create a new feature request for different types of path-based dependencies which *can* be somehow updated

**Comment by bburnichon** [comment]
* minimal reproduction repo with one dependency: https://github.com/bburnichon/renovate-composer-path/tree/minimal-path-invalid-packagist-lookup

**Comment by HonkingGoose** [comment]
> * We need a minimal reproduction repo for a scenario where a path dependency is being looked up from Packagist incorrectly. Presumably it can be as small as one dependency?
> 
> * We mark this issue as status:ready
> 
> * Update code to ensure it skips such dependencies and close this issue
> 
> * Create a new feature request for different types of path-based dependencies which _can_ be somehow updated

Seeing as we have the minimal reproduction, I assume we're safe to move to `status:ready` now. If not please undo. 😉
